parsons/action_kit/action_kit.py

```diff
@@ -463,3 +465,83 @@ def create_generic_action(self, page, email=None, ak_id=None, **kwargs):
+                See the `autocreate_user_fields documentation
+                  <https://roboticdogs.actionkit.com/docs/manual/api/rest/uploads.html#create-a-multipart-post-request>`
+        `Returns`:
+            dict
+                success: whether upload was successful
+                progress_url: an API URL to get progress on upload processing
+                res: requests http response object
+        """
+        import_page = check_env.check('ACTION_KIT_IMPORTPAGE', import_page)
+        if csv_file is None:
+            csv_file = check_env.check('ACTION_KIT_CSVFILE', csv_file)
+
+        # self.conn defaults to JSON, but this has to be form/multi-part....
+        upload_client = self._conn({'accepts': 'application/json'})
+        if isinstance(csv_file, str):
+            csv_file = open(csv_file, 'r')
+
+        res = upload_client.post(
+            self._base_endpoint('upload'),
+            files={'upload': csv_file},
+            data={'page': import_page,
+                  'autocreate_user_fields': int(autocreate_user_fields)})
+        rv = {'res': res,
+              'success': res.status_code == 201,
+              'progress_url': res.headers.get('Location')}
+        return rv
+
+    def bulk_upload_table(self, import_page, table, autocreate_user_fields=0):
+        """
+        Bulk upload a table of new users or user updates.
+        See `ActionKit User Upload Documentation
+             <https://roboticdogs.actionkit.com/docs/manual/api/rest/uploads.html>`
+        Be careful that blank values in columns will overwrite existing data.
+
+        If you get a 500 error, try sending a much smaller file (say, one row),
+        which is more likely to return the proper 400 with a useful error message
+
+        `Args:`
+            import_page: str
+                The page to post the action. The page short name.
+            table: Table Class
+                A Table of user data to bulk upload
+                A user_id or email column is required.
+            autocreate_user_fields: bool
+                When True columns starting with "user_" will be uploaded as user fields.
+                See the `autocreate_user_fields documentation
+                  <https://roboticdogs.actionkit.com/docs/manual/api/rest/uploads.html#create-a-multipart-post-request>`
+        `Returns`:
+            dict
+                success: whether upload was successful
+                progress_url: an API URL to get progress on upload processing
+                res: requests http response object
+        """
+        import_page = check_env.check('ACTION_KIT_IMPORTPAGE', import_page)
+        csv_file = StringIO()
+        outcsv = csv.writer(csv_file)
+        for row in table.table:
+            outcsv.writerow(row)
+        return self.bulk_upload_csv(import_page, StringIO(csv_file.getvalue()),
+                                    autocreate_user_fields=autocreate_user_fields)
```

Comment:
This should work utilizing existing `Table` methods so you don't have to use the `csv` module.

```
return self.bulk_upload_csv(import_page, table.to_csv(), autocreate_user_fields=autocreate_user_fields)
```

Reply:
My objection to this is that it uses the file-system, and I'd rather this be able to stay in-memory.

I'm happy to make a `to_csv_memory` in tofrom.py if you prefer that.